Trim optional white space in CombinedHttpHeaders values

ddossot opened the pull request:
Motivation:

The updated HTTP/1.x RFC allows for header values to be `CSV` and separated by `OWS` [1]. `CombinedHttpHeaders` should remove this `OWS` on insertion.

[1] https://tools.ietf.org/html/rfc7230#section-7

Modification:

`CombinedHttpHeaders` doesn't account for the `OWS` and returns it back to the user as part of the value.

Result:

Fixes #6452

fenik17 commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -244,30 +245,59 @@ public static String simpleClassName(Class<?> clazz) {
+        }
+
+        StringBuilder result = new StringBuilder(length + CSV_NUMBER_ESCAPE_CHARACTERS);
+        boolean quoted = isDoubleQuote(value.charAt(start)) && isDoubleQuote(value.charAt(last)) && length != 1;
```
better make `length != 1` as first condition bcz it has lower cost.

fenik17 replied:
If `quoted==true`, can we increment `start` and decrement `last` to exclude check `i == start || i == last` from the loop?

ddossot replied:
I honestly haven't found a way to make the code better while doing the suggested increment/decrement. There's more involved than just excluding the `i == start || i == last` from the loop, and I have tried to not alter too much the existing code in this PR.

@fenik17 please send a follow-up PR to implement the change you envisioned.

fenik17 replied:
Ok. I'll do it soon. Thanks for an effort!

fenik17 replied:
@ddossot done in #6840

fenik17 commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -469,4 +511,41 @@ public static boolean endsWith(CharSequence s, char c) {
+        return start == 0 && end == length - 1 ? value : value.subSequence(start, end + 1);
+    }
+
+    private static int indexOfFirstNonOwsChar(CharSequence value, int length) {
```
IMHO, make sense add comment that if the value don't contain non-OWS chars, it returns `length`, but not `-1`.

normanmaurer replied:
+1

fenik17 commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -34,6 +34,7 @@
     public static final char LINE_FEED = '\n';
     public static final char CARRIAGE_RETURN = '\r';
     public static final char TAB = '\t';
+    public static final char SPACE = ' ';
```
`' '` -> `0x20`?

normanmaurer replied:
I think `' '` is easier to read...

fenik17 replied:
I just wanted to emphasize specific unicode value.
Not everything can be distinguished in appearance )
Spaces may be different: https://www.cs.tut.fi/~jkorpela/chars/spaces.html

ddossot commented:
@normanmaurer @Scottmitch @fenik17 Sorry for the delay in addressing the comments. I've done a few basic changes and invited @fenik17 to submit a follow-up with the changes he has in mind.

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -16,9 +16,10 @@
 import io.netty.handler.codec.DefaultHeaders;
+import io.netty.handler.codec.Headers;
 import io.netty.handler.codec.ValueConverter;
+import io.netty.util.AsciiString;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.util.AsciiString'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on common/src/main/java/io/netty/util/internal/StringUtil.java:
```diff
@@ -244,30 +245,59 @@ public static String simpleClassName(Class<?> clazz) {
+     *                       according to <a href="https://tools.ietf.org/html/rfc7230#section-7">RFC-7230</a>
+     * @return {@link CharSequence} the escaped value if necessary, or the value unchanged
+     */
+    public static CharSequence escapeCsv(CharSequence value, boolean trimWhiteSpace) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "escapeCsv" is 28 which is greater than 25 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

normanmaurer commented:
@ddossot sorry for the delay :( Just merged into 4.1 (9c1a1916962354f1e69fe43107219c25a86e6cef) . @ddossot would you mind port this over to 4.0 and open a new PR for it as well ?

ddossot commented:
Just for closure here, after discussing with @normanmaurer in meatspace, we will not port to `4.0` as the `CombinedHttpHeaders` feature is lacking altogether there.

normanmaurer commented:
👍

> Am 22.03.2017 um 21:05 schrieb David Dossot <notifications@github.com>:
> 
> Just for closure here, after discussing with @normanmaurer in meatspace, we will not port to 4.0 as the CombinedHttpHeaders feature is lacking altogether there.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
